diff --git a/Source/Core/BoxShadowCache.h b/Source/Core/BoxShadowCache.h
--- a/Source/Core/BoxShadowCache.h
+++ b/Source/Core/BoxShadowCache.h
@@ -11,13 +11,14 @@
 struct BoxShadowGeometryInfo;
 
 struct BoxShadowRenderable : NonCopyMoveable {
-	BoxShadowRenderable(const BoxShadowGeometryInfo& geometry_info);
+	BoxShadowRenderable(RenderManager& render_manager, const BoxShadowGeometryInfo& geometry_info);
 	~BoxShadowRenderable();
 
 	CallbackTexture texture;
 	Geometry geometry;
 	Geometry background_border_geometry;
 	const BoxShadowGeometryInfo& cache_key;
+	RenderManager* render_manager;
 };
 
 class BoxShadowCache {
diff --git a/Source/Core/BoxShadowCache.cpp b/Source/Core/BoxShadowCache.cpp
--- a/Source/Core/BoxShadowCache.cpp
+++ b/Source/Core/BoxShadowCache.cpp
@@ -10,12 +10,14 @@
 namespace Rml {
 
 struct BoxShadowCacheData {
-	StableUnorderedMap<BoxShadowGeometryInfo, WeakPtr<BoxShadowRenderable>> handles;
+	StableUnorderedMap<RenderManager*, StableUnorderedMap<BoxShadowGeometryInfo, WeakPtr<BoxShadowRenderable>>> handles;
 };
 
 static void ReleaseHandle(BoxShadowRenderable* handle);
 
-BoxShadowRenderable::BoxShadowRenderable(const BoxShadowGeometryInfo& geometry_info) : cache_key(geometry_info) {}
+BoxShadowRenderable::BoxShadowRenderable(RenderManager& manager, const BoxShadowGeometryInfo& geometry_info) :
+	cache_key(geometry_info), render_manager(&manager)
+{}
 
 BoxShadowRenderable::~BoxShadowRenderable()
 {
@@ -37,20 +39,21 @@
 static SharedPtr<BoxShadowRenderable> GetOrCreateBoxShadow(RenderManager& render_manager, const BoxShadowGeometryInfo& info)
 {
 	RMLUI_ZoneScoped;
-	auto it_handle = shadow_cache_data->handles.find(info);
-	if (it_handle != shadow_cache_data->handles.end())
+	auto& handles = shadow_cache_data->handles[&render_manager];
+	auto it_handle = handles.find(info);
+	if (it_handle != handles.end())
 	{
 		SharedPtr<BoxShadowRenderable> result = it_handle->second.lock();
 		RMLUI_ASSERTMSG(result, "Failed to lock handle in Box Shadow cache");
 		return result;
 	}
 
-	const auto iterator_inserted = shadow_cache_data->handles.emplace(info, WeakPtr<BoxShadowRenderable>());
+	const auto iterator_inserted = handles.emplace(info, WeakPtr<BoxShadowRenderable>());
 	RMLUI_ASSERTMSG(iterator_inserted.second, "Could not insert entry into the Box Shadow cache handle map, duplicate key.");
 	const BoxShadowGeometryInfo& inserted_key = iterator_inserted.first->first;
 	WeakPtr<BoxShadowRenderable>& inserted_weak_data_pointer = iterator_inserted.first->second;
 
-	auto shadow_handle = MakeShared<BoxShadowRenderable>(inserted_key);
+	auto shadow_handle = MakeShared<BoxShadowRenderable>(render_manager, inserted_key);
 	GeometryBoxShadow::GenerateTexture(shadow_handle->texture, shadow_handle->background_border_geometry, render_manager, inserted_key);
 
 	Mesh mesh;
@@ -67,13 +70,18 @@
 	// There are no longer any users of the cache entry uniquely identified by the handle address. Start from the
 	// tip (i.e. per-color data) and remove that entry from its parent. Move up the cache ancestry and erase any
 	// entries that no longer have any children.
-	auto& handles = shadow_cache_data->handles;
+	auto& managers = shadow_cache_data->handles;
+	auto it_manager = managers.find(handle->render_manager);
+	RMLUI_ASSERT(it_manager != managers.end());
+	auto& handles = it_manager->second;
 	const BoxShadowGeometryInfo& key = handle->cache_key;
 
 	auto it_handle = handles.find(key);
 	RMLUI_ASSERT(it_handle != handles.cend());
 
 	handles.erase(it_handle);
+	if (handles.empty())
+		managers.erase(it_manager);
 }
 
 SharedPtr<BoxShadowRenderable> BoxShadowCache::GetHandle(Element* element, const ComputedValues& computed)
